includes/Checker/Checks/Plugin_Updater_Check.php

```diff
@@ -0,0 +1,226 @@
+	 * @var int
+	 */
+	protected $flags = 0;
+
+	/**
+	 * Constructor.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param int $flags Bitwise flags to control check behavior.
+	 */
+	public function __construct( $flags = self::TYPE_ALL ) {
+		$this->flags = $flags;
+	}
+
+	/**
+	 * Amends the given result by running the check on the given list of files.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 * @param array        $files  List of absolute file paths.
+	 *
+	 * @throws Exception Thrown when the check fails with a critical error (unrelated to any errors detected as part of
+	 *                   the check).
+	 */
+	protected function check_files( Check_Result $result, array $files ) {
+		$php_files = self::filter_files_by_extension( $files, 'php' );
+
+		// Looks for "UpdateURI" in plugin header.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATE_URI_HEADER ) {
+			$this->look_for_update_uri_header( $result );
+		}
+
+		// Looks for special updater file.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATER_FILE ) {
+			$this->look_for_updater_file( $result, $php_files );
+		}
+
+		// Looks for plugin updater code in plugin files.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATERS ) {
+			$this->look_for_plugin_updaters( $result, $php_files );
+		}
+
+		// Looks for plugin updater routines in plugin files.
+		if ( $this->flags & self::TYPE_PLUGIN_UPDATER_ROUTINES ) {
+			$this->look_for_updater_routines( $result, $php_files );
+		}
+	}
+
+	/**
+	 * Looks for UpdateURI in plugin header and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result The check result to amend, including the plugin context to check.
+	 */
+	protected function look_for_update_uri_header( Check_Result $result ) {
+		if ( ! function_exists( 'get_plugin_data' ) ) {
+			require_once ABSPATH . 'wp-admin/includes/plugin.php';
+		}
+
+		$plugin_main_file = WP_PLUGIN_DIR . '/' . $result->plugin()->basename();
+		$plugin_header    = get_plugin_data( $plugin_main_file );
+		if ( ! empty( $plugin_header['UpdateURI'] ) ) {
+			$this->add_result_error_for_file(
+				$result,
+				true,
+				__( 'Plugin Updater detected. Use of the Update URI header is not helpful in plugins hosted on WordPress.org.', 'plugin-check' ),
+				'plugin_updater_detected',
+				$plugin_main_file
+			);
+		}
+	}
+
+	/**
+	 * Looks for plugin updater file and amends the given result with an error if found.
+	 *
+	 * @since n.e.x.t
+	 *
+	 * @param Check_Result $result    The check result to amend, including the plugin context to check.
+	 * @param array        $php_files List of absolute PHP file paths.
+	 */
+	protected function look_for_updater_file( Check_Result $result, array $php_files ) {
+
+		$plugin_update_file = self::filter_files_by_regex( $php_files, '/plugin-update-checker\.php$/' );
+
+		if ( $plugin_update_file ) {
+			foreach ( $plugin_update_file as $file ) {
```

Comment:
Nit-pick: Since it's an array, better to use a plural variable name.
```suggestion
		$plugin_update_files = self::filter_files_by_regex( $php_files, '/plugin-update-checker\.php$/' );

		if ( $plugin_update_files ) {
			foreach ( $plugin_update_files as $file ) {
```